### ISSUE TITLE
C#: The compiler should be consistent about issuing error CS1612 when modifying readonly rvalue struct

### ISSUE THREAD

**Issue by slepmog** [dotnet/roslyn#45284]
**Version Used**: 
VS 16.6.2

**Abstract**
The C# compiler appears to be inconsistent about issuing [CS1612][1] *Cannot modify the return value of 'expression' because it is not a variable*.

When the value being modified is a `readonly struct`, the compiler allows the "mutation" (apparently, on the basis that it cannot occur anyway, because the struct is `readonly`), but only if the struct was returned from a function.
If the struct was returned from a property getter, the compiler issues CS1612, even though the getter-returned value is no more of an rvalue than a function-returned value.

**Steps to Reproduce**:
Have the following code:

```cs
static void Main()
{
    var f = new S1Factory();

    // "Mutating" an rvalue struct returned from a function.
    // The compiler allows that, because no mutation can actually occur
    // due to the struct being readonly, so no reason to raise error CS1612.
    f.GetNewS1AsFunction().CausesSideEffectsWhenWrittenTo = 42;

    // Same, but this time the struct is returned from a property getter.
    // It is just as much of a readonly rvalue as above, but this time the compiler
    // raises CS1612. It shouldn't. Or, it should raise it for the above too.
    f.GetNewS1AsProperty.CausesSideEffectsWhenWrittenTo = 42;
}


readonly struct S1
{
    // Write-only property. Because the struct is readonly,
    // the property only exists for its side effects
    public readonly int CausesSideEffectsWhenWrittenTo {
        set {
            // Cause side effects. The struct cannot be mutated though
            // because it's readonly.
        }
    }
}

readonly struct S1Factory
{
    public readonly S1 GetNewS1AsProperty => new S1();
    public readonly S1 GetNewS1AsFunction() => new S1();
}
```

**Expected Behavior**:
Both assignment lines in the `Main()` compile.
Or, both don't compile - I am honestly not sure which one is correct.

**Actual Behavior**:
The line with `f.GetNewS1AsFunction()` compiles, the line with `f.GetNewS1AsProperty` does not, which is inconsistent and confusing.

Note that if you remove the `readonly` modifier from `struct S1`, then the compiler will correctly issue CS1612 on both assignment lines, because now that the struct is no more `readonly`, a call to the write-only property is allowed to mutate the struct, and because it's an rvalue, the error is correctly raised.

Original StackOverflow question: https://stackoverflow.com/q/62444352/

[1]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1612

**Comment by carlreinke** [comment]
I'd prefer that both compile.  One example use case is implementing fake indexed properties:

```C#
public sealed class Example
{
    private readonly int[] _indexedPropertyValues = new int[1];

    public IndexedPropertyValues IndexedProperty => new IndexedPropertyValues(this);

    public IndexedPropertyValues GetIndexedProperty() => new IndexedPropertyValues(this);

    public ref int IndexedProperty2(int index) => ref _indexedPropertyValues[index];

    public readonly struct IndexedPropertyValues
    {
        private readonly Example _this;

        internal IndexedPropertyValues(Example specializedDictionary)
        {
            _this = specializedDictionary;
        }

        public int this[int index]
        {
            get => _this._indexedPropertyValues[index];
            set => _this._indexedPropertyValues[index] = value;
        }
    }
}

public static class Program
{
    public static void Main(string[] args)
    {
        var example = new Example();

        example.IndexedProperty[0] = 1;  // CS1612

        example.GetIndexedProperty()[0] = 1;  // Okay (but ugly)

        example.IndexedProperty2(0) = 1;  // Okay (but ugly)
    }
}
```

**Comment by CyrusNajmabadi** [comment]
Wouldn't it make more sense to write your property as:

```c#
        public ref int this[int index] => ref _this._indexedPropertyValues[index];
```

?

**Comment by carlreinke** [comment]
That only works if you only have one.  Also if the indexed property is backed by a growable array, you might not want to hand out refs.

**Comment by CyrusNajmabadi** [comment]
In that case, using a method seems sensible here to me.

**Comment by carlreinke** [comment]
Today, yeah, you either give up the indexer syntax and use a method or you give up being allocation-free and make the fake indexed property return a reference type.

Given that the problem CS1612 is meant to protect against (mutating the copy of the struct returned by the property) can't happen when the returned struct is readonly, it seems reasonable for the compiler to not produce the error.

**Comment by carlreinke** [comment]
There was another alternative that I missed: return a ref from the property.

```C#
public sealed class Example
{
    private readonly int[] _indexedPropertyValues = new int[1];

    private readonly IndexedPropertyValues _indexedProperty;

    public Example()
    {
        _indexedProperty = new IndexedPropertyValues(this);
    }

    public ref readonly IndexedPropertyValues IndexedProperty => ref _indexedProperty;

    public readonly struct IndexedPropertyValues
    {
        private readonly Example _this;

        internal IndexedPropertyValues(Example @this)
        {
            _this = @this;
        }

        public int this[int index]
        {
            get => _this._indexedPropertyValues[index];
            set => _this._indexedPropertyValues[index] = value;
        }
    }
}

public static class Program
{
    public static void Main()
    {
        var example = new Example();

        example.IndexedProperty[0] = 1;
    }
}
```

This seems to me like the least-bad way to work around CS1612 for the fake-indexed-property use case.

**Comment by MrBrixican** [comment]
I'd also like to see both cases compile. The struct still remains immutable for all intents and purposes, so there is nothing gained by prohibiting this functionality. This enables a prime use case in easily revocable, immutable, garbage free handles for an object pooling management system, specifically in relation to high performance applications like games.

**Comment by CyrusNajmabadi** [comment]
@MrBrixican can you use methods in that case?

**Comment by MrBrixican** [comment]
@CyrusNajmabadi  Ah, you are correct. I meant to say _more easily enables_. Properties are far more convenient wrappers around the typical getter/setter method combo (especially for +=, -=, etc) and significantly reduce the api surface a consumer would need to understand. Of course, methods serve the role well enough in the interim, but this issue should still be addressed seeing as properties are foundational to the C# language.

**Comment by Chicken-Bones** [comment]
I'd like to bump this one in the hopes of getting it out of backlog. I publish an API for a large community of users and this issue creates friction.

Simplified
```cs
public readonly struct Tile
{
	private readonly uint _index;
	public byte TileColor { get => ...; set => ...; }
	...
}

// elsewhere
public readonly struct Tilemap
{
	public Tile this[int x, int y] => ...
}

// Cannot write
Tiles[x, y].TileColor = 0;

// Have to write
Tile t = Tiles[x, y]:
t.TileColor = 0;
```

In other places we have been able to use `ref` returns with get only properties, but here we need an actual property.

**Comment by MrBrixican** [comment]
While I appreciate that a _workaround_ is being offered, this issue was created with the basis that something that should be working, readonly struct pass-through properties that _do not_ violate the immutable contract, does not work. Whether a design pattern is "abuse" or not, is not relevant. C# is a tool, and it is up to developers themselves to not misuse it.

Properties are a strength of C# and are designed with mutation in mind, hence setters existing. The sentiment that only methods should be used for mutation is something I'd expect more from a JVM or Javascript discussion, not dotnet, haha.

**Comment by CyrusNajmabadi** [comment]
> Whether a design pattern is "abuse" or not, is not relevant.

It is very relevant.  We've designed the language, and compiler rules, around the patterns we think are appropriate and which we think are not.  That something *could* possibly be done still has to fall to the question of *should* it be done.

**Comment by Chicken-Bones** [comment]
In our case, the `struct` being `readonly` is a performance detail for the jit, and the properties accessing an external (structure of arrays) data store is an implementation detail not relevant to the user.

Users are not unfamiliar with structs being references to actual data (such as `Memory`)

**Comment by CyrusNajmabadi** [comment]
> In our case, the struct being readonly is a performance detail for the jit and compiler, 

'readonly' means that non-readonly methods will make copies.  If you don't make your struct readonly, then you have no problem, as the compiler will glad mutate something non-readonly.

'readonly' isn't for performance.  It's to actually ensure that it can't mutate.  

>  and the properties accessing an external (structure of arrays) data store is an implementation detail

But it has to not be an impl detail.  Because the only way this could work would be if the property doesn't mutate the value itself.  And a setter which isn't actually mutating the thing you're calling on *is* the anti-pattern, and is much more cleanly and clearly expressed as a method call.

**Comment by Chicken-Bones** [comment]
Correct `readonly` prevents defensive copies being created which would serve us no purpose.

Originally, `Tile` was a `class`, and `Tiles[x, y]` always returned a reference to the same class. We had two issues with this approach, but performance and memory overhead was a primary concern. In an update ~2 years ago we changed `Tile` to a `struct` containing a single `_index` and moved the data into arrays. This eliminated object overheads, and improved cache coherency. 
`readonly struct Tile` is effectively a validated reference, or a smart pointer, etc. 

Users understand that `Tile` being a struct does not mean the effects of their setters are discarded. They can continue to write code as if `Tile` was a `class`. Moving them to `GetColor()` `SetColor(Color color)` method pairs provides no additional readability or clarity.

I understand that seeing `c.Color = value` where `c` is a local copy of a `struct` raises some questions, but I believe that seeing `c.SetColor(value)` raises the exact same questions, and that these questions are quickly addressed by documentation.

I understand that we need to consider some subset of possible ways to write C# to be 'good/proper/intended', so I respect the time you're spending making this decision.

In my experience, one of the strengths of C# is that it provides tools to obtain better performance and implement closer to the metal, while retaining the safety of a managed language. I've always seen `struct` as primarily performance and native interop motivated, and recent focus on `ref` fields indicates that this is an opinion shared by the language team.

**Comment by CyrusNajmabadi** [comment]
>  but I believe that seeing c.SetColor(value) raises the exact same questions

It raises no questions for me.  As methods are in the space of doing anything desired, whereas properties very much are designed such that they can be thought of as fields, just with computation being allowed on them.  

Can you just expose this as a ref?

**Comment by carlreinke** [comment]
> And a setter which isn't actually mutating the thing you're calling on _is_ the anti-pattern

Implementing a mutable view using a `readonly struct` is not an anti-pattern.  `Span<T>` is a `readonly` `struct`, and yet mutating something other than the `struct` itself is the reason that it exists!

**Comment by CyrusNajmabadi** [comment]
@carlreinke What mutable properties does Span have that you get CS1612 with?

**Comment by carlreinke** [comment]
@CyrusNajmabadi It seems you entirely missed the point.

**Comment by CyrusNajmabadi** [comment]
@carlreinke You seemed to be pointing at Span as an example of why something should be allowed.  But Span itself doesn't seem to have any of the properties (no pun intended) that make it a suitable analogy for what's going on here.

**Comment by carlreinke** [comment]
I am pointing at `Span<T>` as an example of a mutable view implemented using a `readonly` `struct`, which is what you claimed is an anti-pattern.  The only difference between what `Span<T>` does and what CS1612 is preventing here is that `Span<T>` uses an indexer (which is an indexed property) whereas CS1612 prevents using a non-indexed property for the same effect.

**Comment by CyrusNajmabadi** [comment]
> I am pointing at Span<T> as an example of a mutable view implemented using a readonly struct, which is what you claimed is an anti-pattern.

That's not what i claimed.  What i said was:

> And a setter which isn't actually mutating the thing you're calling on is the anti-pattern

and

> As methods are in the space of doing anything desired, whereas properties very much are designed such that they can be thought of as fields, just with computation being allowed on them.

'Span' doesn't change things for me as 'Span' isn't doing anything off wrt statements.

**Comment by CyrusNajmabadi** [comment]
> whereas CS1612 prevents using a non-indexed property for the same effect.

Correct.  As i asked before: Can you just expose this as a ref?

That's what `Span` does here.  If you want to be more like `Span` that's totally reasonable to me.  An explicit `ref` access is def a way to indicate that you behave in this fashion.

**Comment by Chicken-Bones** [comment]
> > but I believe that seeing c.SetColor(value) raises the exact same questions
> 
> It raises no questions for me. As methods are in the space of doing anything desired, whereas properties very much are designed such that they can be thought of as fields, just with computation being allowed on them.
> 
> Can you just expose this as a ref?

We have done so for other members, but some of them require a property to perform bitpacking. Whether a property getter returns a `ref` or whether the property has a `set` method is only identifiable by interrogating the underlying member, not from reading code in plain text. The use cases are often similar, though it remains that neither refs nor properties are first class language citizens yet.

I understand your stance on properties behaving as fields rather than get/set methods. I think that allowing setters at all in `readonly` structs contradicts this, but nonetheless. 

I believe further agreement could only be obtained by consulting more perspectives, as C# caters to a diverse set of users with different expectations and use cases.

**Comment by carlreinke** [comment]
@CyrusNajmabadi  Okay, fair enough, with `Span<T>` there is a `ref` along the way, but the API user doesn't see the `ref` in their code.  Assuming you can build an API that using `ref` that's identical in usage to the API you wanted to build, what's the value that CS1612 is providing here?  All it did was require you have something to `ref` that you otherwise might not have needed.

**Comment by CyrusNajmabadi** [comment]
> All it did was require you have something to ref that you otherwise might not have needed.

To me... that is the value :)  

It strongly pushes one down the path of only exposing in this fashion through either a method or a ref-property.  Normal props are blocked off.

**Comment by dmitry-azaraev** [comment]
Generally ref-struct never should be used, unless you really have reasons and want to. When I'm meet this issue in 2020 - i'm owned in hands two readonly structs, which are copyable and valid all their lifetime, more over they always constructed on the fly, so they never directly stored anywhere.

There is impossible to put any sensible argue because language is just a tool, and it should prevent only technically invalid programs. It MUST not prevent to write valid programs. C# with CS1612 doing exactly this.

**Comment by MrBrixican** [comment]
I think we need to review some statements in this issue so far:

> As methods are in the space of doing anything desired, whereas properties very much are designed such that they can be thought of as fields, just with computation being allowed on them.

This is not in line with the official C# language spec ([§15.7.1](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1571-general)):
> However, unlike fields, properties do not denote storage locations. Instead, properties have accessors that specify the statements to be executed when their values are read or written. Properties thus provide a mechanism for associating **actions** with the reading and writing of an object’s characteristics;

Clearly, actions are very much included in the scope of property setters. Many dotnet classes make use of these "side-effects" for events/notifications. Look no further than some of the most used classes in enterprise settings: `DataTable`, `DataRowCollection`, and `DataRow` ([source](https://source.dot.net/#System.Data.Common/System/Data/DataTable.cs,4b6eb411ec312017,references)).

It has also been mentioned that a property setter that does not directly modify the item you're calling it on is an anti-pattern:
> And a setter which isn't actually mutating the thing you're calling on is the anti-pattern, and is much more cleanly and clearly expressed as a method call.

On top of already contradicting §15.7.1 regarding no standardized storage location, this also contradicts with more Microsoft code:
- `FileStream.Position` setter redirects to `FileStreamStrategy.Position` ([source](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs,e23a38af5d11ddd3))
- `Console` uses `ConsolePal` in setters and getters ([source](https://source.dot.net/#System.Console/System/Console.cs,76c928bc19a5b758))

All of these were found after only a few minutes of searching, there are sure to be countless more contradictions.

---

All that is to say, code correctness is so subjective, it is not even worth discussing in this context. Let's please stay on topic, this issue is not about _good_ code, anti-patterns, or best practices. This is about an inconsistency in the way readonly struct property setters are treated by the compiler:

```
internal class Program
{
    ReadonlyStruct PropertyReadonlyStruct { get; set; }
    ReadonlyStruct FieldReadonlyStruct;
    ReadonlyStruct GetReadonlyStruct() => PropertyReadonlyStruct;
    ReadonlyStruct this[int index] => PropertyReadonlyStruct;

    void Main()
    {
        FieldReadonlyStruct.Data = 0;     // Case 1: no issue
        PropertyReadonlyStruct.Data = 0;  // Case 2: CS1612
        GetReadonlyStruct().Data = 0;     // Case 3: no issue
        this[0].Data = 0;                 // Case 4: CS1612
    }

    public readonly struct ReadonlyStruct
    {
        public int Data { get => 0; set { /* no-op */ } } // Note no mutation of ReadonlyStruct instance
    }
}
```

Cases 1 and 3 currently work, and given that breaking changes are generally avoided they will continue to work in future. It only makes logical sense then, to maintain consistency, cases 2 and 4 should also work. 

Regardless of my own bias, I think it's pretty obvious that removing this error in these cases will at the very least reduce confusion among developers which we've seen for the last 4 years.

However, I'm not on the boat of **"fix it right now"** and understand that the implementation to address this issue may be far more complex than many of us imagine. I just think it would be neat if this issue could be given a little more consideration.  :)

**Comment by CyrusNajmabadi** [comment]
> On top of already contradicting §15.7.1 regarding no standardized storage location, this also contradicts with more Microsoft code:

These do not contradict what i was saying.  These are not readonly.  

> All that is to say, code correctness is so subjective, it is not even worth discussing in this context

It is very worth discussing.  Stating that you're unwilling to consider a core concern that is one of the reasons we have this behavior in the first place isn't beneficial to forward progress here.

**Comment by CyrusNajmabadi** [comment]
> Cases 1 and 3 currently work,

Correct.  The rules here model the sorts of api patterns that would be viable and appropriate in these cases.  It's very intentional that hte cases that do not work have that behavior, and steering APIs away from working this way was intentional.

**Comment by CyrusNajmabadi** [comment]
> There is impossible to put any sensible argue because language is just a tool, and it should prevent only technically invalid programs. It MUST not prevent to write valid programs. C# with CS1612 doing exactly this, e.g. prevent to write valid programs.

That's not how language design works for C# (or virtually any other language for that matter).  And the language has *many* rules that are about blocking things that we technically *could* allow, but do not feel are good to allow.

For example, we do not allow 'collection initializers' on types that do not implement `IEnumerable`.  We also don't allow you to write code like the following:

```c#
a + b; // Only assignment, call, increment, decrement, await, and new object expressions can be used as a statement
```

Yes, we could relax all these rules and allow this to be compiled, for the rare user that wants to do something like this.  But that's not our philosophy.  We will often block by default if we feel like this just isn't a good pattern.  And if someone wants this to be legal, the onus is on the person wanting that to make a *very strong argument* that will convince us on hte lang design team to make this legal. :)

**Comment by MrBrixican** [comment]
> These do not contradict what i was saying. These are not readonly.

However, it does contradict. [§16 Structs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/structs) gives no specifics on _what_ a property is, thus I'm going to assume the intention is to fall back to [§15.7.1](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1571-general) as per [§16.3](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/structs#163-struct-members).

Interestingly, §16 does have a few addendums regarding struct properties, though:

[§16.4.12.6](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/structs#164126-method-and-property-invocation) describes, in Microsoft's own words, that a getter or setter is, in fact, treated as a method call. 
>A value resulting from a method invocation e1.M(e2, ...) or property invocation e.P has safe-context of the smallest of the following contexts:
> - caller-context.
> - The safe-context of all argument expressions (including the receiver).

> A property invocation (either get or set) is treated as a method invocation of the underlying method by the above rules.

[§16.2.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/structs#1622-struct-modifiers) describes that readonly _instance_ properties can't have a setter declaration. Either this rule is not being enforced since setters that respect immutability are allowed, or more likely, non-auto-property setters in such cases are not considered _instance_ properties, but instead just convenient wrappers around a method body as suggested above:
> A readonly struct has the following constraints:
> - Each of its instance fields shall also be declared readonly.
> - None of its instance properties shall have a set_accessor_declaration ([§15.7.3](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1573-accessors)).
> - It shall not declare any field-like events ([§15.8.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1582-field-like-events)).

From the above we can conclude the following:
- Non-auto-property setters have no enforced or implied implementation regarding storage location
- Struct property invocations are synonymous with method invocations
- readonly struct non-auto-properties are just convenient wrappers around their method bodies

Thus, the distinction between methods and properties in this context is not aligned with language spec, and ultimately irrelevant.

---

As an aside, let's take a look at the CS1612 message:
> Cannot modify the return value of 'expression' because it is not a variable

In my previous examples, was the struct ever "modified"? No. Byte for byte regardless of what you set the property to, the struct is exactly the same, as enforced by readonly. The error does not apply.

If there is a desire to enforce patterns, please create a new honest error with the following:
> Cannot call set accessor on readonly struct property 'property' when returned from accessor, in order to prevent a perceived anti-pattern, not due to infeasibility. Please assign the struct to a local variable or field to ignore this guard entirely.

I offer this to you under MIT so feel free to use it. :)

---

As others have noted, this change is not something that affects the average developer, as struct implementations are rarely explicitly defined in the majority of C# enterprise code, readonly even less so. This is for advanced, niche, high performance use cases where it can be assumed the developer a) knows what they are doing and b) wants to offer a less off-putting/clunky user experience to api consumers.

In a language that allows operator overloading (particularly implicit operator) and unsafe code, there has obviously been a spirit of giving the developer the tools they need to do what they want with minimal barriers. The infantilization of the developer that is being pushed seems very contrary to what has made C# popular in the first place.

**Comment by CyrusNajmabadi** [comment]
> escribes, in Microsoft's own words, that a getter or setter is, in fact, treated as a method call.

Yes.  For the purposes of safe-context determination.  That's not related to this convo though.

**Comment by CyrusNajmabadi** [comment]
> None of its instance properties shall have a set_accessor_declaration ([§15.7.3](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1573-accessors)).

It looks like that is stale.  @BillWagner for thoughts on fixing that up.

**Comment by CyrusNajmabadi** [comment]
> In my previous examples, was the struct ever "modified"? No. 

And that's why this is an antipattern.  The code certainly gives that appearance.  If you're going to have a type that behaves in that way (where it doesn't seem like it will mutate, but the code indicates it will), just don't use a non-ref property for that.  :)

>Non-auto-property setters have no enforced or implied implementation regarding storage location

That's why i said that this wasn't a question of whether something *could* be allowed, but rather *should* it be allowed.  This falls into the latter discussion.  I completely understand the arguments of making this just about "could this be allowed", but they're falling flat for me because (like other examples provided in this thread) i think the additional guardrails are good things.  Indeed, the examples given so far have *reinforced* that belief for me due to how wonky they actually appear.  

I think this thread is lacking in real world examples that demonstrate why this change would be good.  So far, all the examples have made me actually feel like we don't go *far enough* in terms of blocking certain things.

**Comment by BillWagner** [comment]
> > None of its instance properties shall have a set_accessor_declaration ([§15.7.3](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#1573-accessors)).
> 
> It looks like that is stale. @BillWagner for thoughts on fixing that up.

The ECMA spec is out of date. The committee is working on updating it. 

Right now, the ECMA committee is working on the spec for C# 8. We haven't merged any of the initial PRs for those features. That's why the C# 8 speclets are still published. `init` only setters were added in C# 9. The init-only setters spec details what will change for [readonly structs](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/init#readonly-structs). 

I admit this isn't ideal. We're continuing to work to catch up. But a long hiatus, and a (mostly) volunteer committee means progress takes time.

**Comment by MrBrixican** [comment]
> And that's why this is an antipattern. The code certainly gives that appearance. If you're going to have a type that behaves in that way (where it doesn't seem like it will mutate, but the code indicates it will), just don't use a non-ref property for that. :)

> I think this thread is lacking in real world examples that demonstrate why this change would be good. So far, all the examples have made me actually feel like we don't go far enough in terms of blocking certain things.

@Chicken-Bones has already given a good example as to why ref properties would not work, bit packing. Where the binary representation of data is not 1 to 1 with how the user should have to interact with it.

I have given the example of non garbage collected immutable handles that set and get values through interop, where you would clearly not want the value to be stored in the struct or have the ability to set the values directly via ref in the first place. Why would you not want to interact with verbose C getters and setters through beautiful, concise C# properties? Can you honestly say, as an api consumer, you would prefer methods over properties? [Here's](https://github.com/MrBrixican/Foster.Audio/blob/main/Foster.Audio/SoundInstance.cs) a more concrete example in an audio library I've been tinkering with. As an example, moving a sound in 3d space would go from:
```
this.Music.Position += new Vector3(1f);
```
to
```
this.Music.SetPosition(this.Music.GetPosition() + new Vector3(1f));
```
If you really think that methods are a better experience, either you're joking or you may find [this](https://www.oracle.com/java/technologies/javase/javacareersatoracle.html) link useful.

I'm not sure if we've not been clear enough or perhaps we need the input of Microsoft developers with more diverse backgrounds in development.

**Comment by CyrusNajmabadi** [comment]
> If you really think that methods are a better experience, either you're joking

I'm not joking.

**Comment by CyrusNajmabadi** [comment]
> Can you honestly say, as an api consumer, you would prefer methods over properties?

Yes, when dealing with readonly types and non-locations rvalues.  But, personally, i wouldn't make these readonly.

**Comment by Chicken-Bones** [comment]
> Yes, when dealing with readonly types and non-locations rvalues. But, personally, i wouldn't make these readonly.

This depends on what the purpose of the `readonly` modifier on types is. As it stands, there are multiple purposes which aren't fully aligned, the result of which is friction here. 
1. Preventing the modification of the values of fields in the type. Providing a guard rail for developers working on the type.
2. Signalling to consumers of the type, that none of the properties or methods on the type can modify the values of its fields.
3. Allowing the compiler to eliminate defensive copies.

Purposes 2 and 3 are both derived from the core functional specification, but they come at odds here, otherwise I would be okay to remove `readonly` modifier from my struct.

Regarding purpose 2, the convention:
- I expect many consumers of `struct` types will check the definition to see the `readonly` modifier. 
- Unlike `ref` it has no effect or restriction on the behaviour of user code (outside of this warning). It can only be used to help reason about the underlying implementation of the type. Personally, I like to know the implementation of my types, not just their interface, but that's because I'm a low level programmer, not one who 'programs against the interface'
- Property setters on `readonly` types appear to initially pose a conflict, as you've described. 
"How can a readonly type have a setter which appears to modify it? Properties are like fields, and the fields can't be assigned to." 
As property setters in `readonly` types are not forbidden outright, the user may come to one of two conclusions:
  - The property setter mutates some other data, which is referenced by but not literally held in the struct (in a similar manner to assigning to a `ref` field)
  - The property setter is unsafe/erroneous
  - or they may just remain confused.
  
As you've said, it remains the burden of the proposal champions to make a compelling use-case for opening up of a restriction. 

I believe this restriction only really affects developers who are trying to make 'pointer-like' structs and are also concerned with performance because their structs are 'dereferenced' repeatedly in hot loops. This is exactly the ECS game development pattern which has risen in popularity in recent years. You can see other developers with the same scenario and motivation in https://github.com/dotnet/csharplang/discussions/2068

Additional notes:
- Even if the compiler didn't eliminate defensive copies, the `readonly` modifier still has safety benefit to type authors in ensuring they don't accidentally try and mutate the struct when maintaining or adding members, properties notwithstanding.
- If the 'extension everything' or 'roles' proposals are eventually implemented, there would be further motivation for revisiting this error. Perhaps extension setters would be forbidden on `readonly` types, but the ECS or other 'entity key struct' use cases would benefit heavily from allowing module authors to define extension properties on entities which access their additional component data types.

**Comment by CyrusNajmabadi** [comment]
> I believe this restriction only really affects developers who are trying to make 'pointer-like' structs and are also concerned with performance because their structs are 'dereferenced' repeatedly in hot loops.

IMO, if you have such a situation, writing methods seems entirely cromulent :)

The benefit here is that if you have the normal property form the code looks entirely suspect/broken.  In the examples, since a struct is returned, normal assumption is that that would be a copy.  And so, mutating that copy is more often than not just a straight bug.  Akin to getting a tuple as an rvalue and writing into it.  

Methods much more clearly express: do unbounded work, including on data *outside* of the value.  Properties that "call through" to some other store are the much less common case, and it seems fine to restrict those to catch the common case of this being a bug.   Methods, on the other hand, allow you to express that something special can be going on, and have hte API stay aligned with the intent and clarity of the impl.

**Comment by Chryssie** [comment]
> Methods much more clearly express: do unbounded work, including on data outside of the value.
When talking about the rvalue though, the method is done and the impl of that method doesn't have a say on whether or not mutating the rvalue is valid or not.

That is unless if mutating the struct would not require a copy. Which, to that, it's worth noting that calling a `readonly` setter on a `readonly` field compiles whereas calling a non-`readonly` setter on a `readonly` field gives "CS1648: Members of a readonly field 'Program.ReadOnlyField' cannot be modified (except in a constructor or a variable initializer)".

```cs
class Program
{
    void Main()
    {
        ReadOnlyField.ReadOnlyData = default; // No error
        Property.ReadOnlyData = default; // CS1612: Cannot modify the return value of 'Program.Property' because it is not a variable.
        Method().ReadOnlyData = default;

        ReadOnlyField.DataProperty = default; // CS1648: Members of a readonly field 'Program.ReadOnlyField' cannot be modified (except in a constructor or a variable initializer)
        Property.DataProperty = default; // CS1612: Cannot modify the return value of 'Program.Property' because it is not a variable.
        Method().ReadOnlyData = default; // No error
    }

    readonly StructWithAField ReadOnlyField;
    StructWithAField Property => default;
    StructWithAField Method() => default;

    struct StructWithAField
    {
        public int DataProperty
        {
            get => default;
            set { }
        }
        public readonly int ReadOnlyData
        {
            get => default;
            set { }
        }
    }
}
```

In that case the error is different, `CS1648` as oppose to `CS1612`, however, I don't think that impacts whether or not the code itself looks broken.

**Comment by CyrusNajmabadi** [comment]
/shrug :)   Which errors get reported aren't really that important to me (though consistency across the scenarios would certainly be nice).  What matters is the semantics we want to block and what we want to allow.

**Comment by Xenoprimate** [comment]
Bumping this topic to add that I'm encountering the same inconsistency and it is affecting my ability to provide a concise/consistent library API. I've read the comment chain above and don't have much extra to add except that I agree with those commenting in favour of this change.

I _will_ say that I find it incongruent that methods are excepted from this compiler rule but properties are not. I think most C# developers understand properties to basically be a terser method syntax and that properties are "really" methods under-the-hood. Even if there are philosophical differences, I'm personally much more interested in the practicalities; from a pragmatic point of view this restriction as it stands today doesn't "save" anyone from any actual lost mutations and only serves to make a specific API design pattern more inconsistent.

**Comment by CyrusNajmabadi** [comment]
> I will say that I find it incongruent that methods are excepted from this compiler rule but properties are not. 

Petitioning to change the language is required here. The compiler just implements the rules we language designers think are appropriate for the language.

**Comment by CyrusNajmabadi** [comment]
> and only serves to make a specific API design pattern more difficult to use.

That's a virtue if the particular design pattern is not viewed positively. See the above conversations for alternate approaches we think should be used instead.

**Comment by DoktorDemento** [comment]
> This truly feels like an abuse of structs, indexers, and properties, as far as .Net and C# want them to be considered.

If the intention is to discourage property setters on `readonly struct` types, surely it would make more sense to show the warning (or error) _at the property definition_, rather than showing an error when the property is used? The 'crime' has already been committed!

This would also allow developers who feel this restriction is unnecessary (see upthread) to suppress the warning if it would be a false-positive in their case.

**Comment by Xenoprimate** [comment]
> Petitioning to change the language is required here. The compiler just implements the rules we language designers think are appropriate for the language.

There is a concurrent discussion here [https://github.com/dotnet/csharplang/discussions/2068](https://github.com/dotnet/csharplang/discussions/2068); I chose to comment here however as there seemed to be more debate/discussion on this thread. Whether this warning/error change would be a change to the language or the compiler I'm not personally sure of, so I simply picked the more active thread.

> That's a virtue if the particular design pattern is not viewed positively. See the above conversations for alternate approaches we think should be used instead.

I'm aware of the alternatives proposed but agree with the other commenters that they're inadequate. [MrBrixican's comment above](https://github.com/dotnet/roslyn/issues/45284#issuecomment-2021091037) applies to my scenario as well (non-GCd immutable handles). The fact that they're non-GCd immutable handles is not a detail I want users to have to know or care about. I wish to provide the illusion of regular mutable C# objects. However, this compiler error restricts that, forcing me to push a performance-oriented implementation detail on to the end-user.

Setter methods don't address the concerns raised anyway. A "SetXyz(p)" method with void return and no other inputs by appearance alone still ostensibly mutates the instance; understanding that it does not requires an implicit understanding that the target struct is a convenience object wrapping a handle-- therefore the distinction between setter and set property is dubious at best. Consequently, I consider that alternative to be forcing an uglier anti-conventional pattern for no actual gain.

I would prefer CS1612 be there to help protect users against accidentally attempting to mutate rvalues, not to make judgement calls about the design of my API/library.

**Comment by jnm2** [comment]
Per @jaredpar, changing this now requires a discussion with a minispec. Championed: https://github.com/dotnet/csharplang/issues/9174
